packages/remix-react/routes.tsx

```diff
@@ -198,40 +217,104 @@ export function createClientRoutes(
             )
           : routeModule.shouldRevalidate,
       });
-    } else {
-      // Load all other modules via route.lazy()
-      Object.assign(dataRoute, {
-        ...dataRoute,
-        lazy: async () => {
-          let mod = await loadRouteModuleWithBlockingLinks(
-            route,
-            routeModulesCache
-          );
-
-          let lazyRoute: Partial<DataRouteObject> = { ...mod };
 
-          if (needsRevalidation) {
-            lazyRoute.shouldRevalidate = wrapShouldRevalidateForHdr(
-              route.id,
-              mod.shouldRevalidate,
-              needsRevalidation
-            );
+      let initialData =
+        initialState &&
+        initialState.loaderData &&
+        initialState.loaderData[route.id];
+
+      let isHydrationRequest =
+        needsRevalidation == null &&
+        routeModule.clientLoader != null &&
+        routeModule.clientLoader.hydrate === true;
+
+      dataRoute.loader = ({ request, params }: LoaderFunctionArgs) => {
+        return prefetchStylesAndCallHandler(async () => {
+          if (!routeModule.clientLoader) {
+            // Call the server when no client loader exists
+            return fetchServerLoader(request);
           }
 
-          return {
-            hasErrorBoundary: lazyRoute.hasErrorBoundary,
-            shouldRevalidate: lazyRoute.shouldRevalidate,
-            handle: lazyRoute.handle,
-            Component: lazyRoute.Component,
-            ErrorBoundary: lazyRoute.ErrorBoundary,
-          };
-        },
-      });
+          return routeModule.clientLoader({
+            request,
+            params,
+            async serverLoader() {
+              if (isHydrationRequest) {
+                isHydrationRequest = false;
+
+                // Throw an error if a clientLoader tries to call a serverLoader that doesn't exist
+                if (initialData === undefined) {
+                  throw new Error(
```

Comment:
Is this something we want to throw on or something we want to just return null / undefined as if there wasn't a loader? Reason this stands out to me is because it makes it seem like implementing shared behavior could become more difficult / leak implementation details.

If I wanted to have a route template that ensured a clientLoader that dealt with a global caching strategy was exported by default from all new routes, they would also have to have a loader that just returned null or the route wouldn't work until one was added. 

Does that sound right?

Reply:
I considered this too and almost removed it since right now we just return `null` in `fetchServerLoader`.  The thing that felt maybe _slightly_ different was in the current case, it's the framework calling it and we _have_ to have a loader to handle the JS/CSS module loads.  In the new flow, it's the user calling a function erroneously so it felt maybe more confusing to let it fail silently.

I could absolutely be convinced to remove this error path though - curious what @mjackson and @ryanflorence think

Reply:
IMO if I call a route serverLoader but I don't have one, it should throw so I either catch it or the error bubbles and got reported to my Sentry

Reply:
yeah it's essentially the same as if you manually did `fetch('/route-without-loader?_data=routes%2Froute-without-loader')` which would throw

Reply:
err....scratch that.  That doesn't _throw_ but returns a 400 with `You made a GET request to "/" but did not provide a `loader` for route "routes/_index", so there is no way to handle the request`.

_Insert debate here on whether fetch should throw on non-200 responses 😂 _